reana_client/validation/environments.py

```diff
@@ -0,0 +1,470 @@
+                            full_image, msg
+                        ),
+                    }
+                )
+                return False
+            else:
+                self.messages.append(
+                    {
+                        "type": "warning",
+                        "message": "==> WARNING: Existence of environment image {} in Docker Hub could not be verified. Status code: {} {}".format(
+                            full_image, response.status_code, response.reason,
+                        ),
+                    }
+                )
+                return False
+        else:
+            self.messages.append(
+                {
+                    "type": "success",
+                    "message": "Environment image {} exists in Docker Hub.".format(
+                        full_image
+                    ),
+                }
+            )
+            return True
+
+    def _get_image_uid_gids(self, image, tag):
+        """Obtain environment image UID and GIDs.
+
+        :returns: A tuple with UID and GIDs.
+        """
+        # Check if docker is installed.
+        run_command("docker version", display=False, return_output=True)
+        # Run ``id``` command inside the container.
+        uid_gid_output = run_command(
+            'docker run -i -t --rm {} sh -c "/usr/bin/id -u && /usr/bin/id -G"'.format(
+                self._get_full_image_name(image, tag)
+            ),
+            display=False,
+            return_output=True,
+        )
+        ids = uid_gid_output.splitlines()
+        uid, gids = (
+            int(ids[-2]),
+            [int(gid) for gid in ids[-1].split()],
+        )
+        return uid, gids
+
+    def _get_full_image_name(self, image, tag=None):
+        """Return full image name with tag if is passed."""
+        return "{}{}".format(image, ":{}".format(tag) if tag else "")
+
+    def _get_local_docker_images(self):
+        """Return a list with local docker images."""
+        # Check if docker is installed.
+        run_command("docker version", display=False, return_output=True)
+        docker_images = run_command(
+            'docker images --format "{{ .Repository }}:{{ .Tag }}"',
+            display=False,
+            return_output=True,
+        )
+        return docker_images.splitlines()
+
+
+class SerialEnvironmentValidator(EnvironmentValidatorBase):
+    """REANA serial workflow environments validation."""
+
+    def validate_environment(self):
+        """Validate environments in REANA serial workflow."""
+        for step in self.workflow_steps:
+            image = step["environment"]
+            kubernetes_uid = step.get("kubernetes_uid")
+            self._validate_environment_image(image, kubernetes_uid=kubernetes_uid)
+
+
+class YadageEnvironmentValidator(EnvironmentValidatorBase):
+    """REANA yadage workflow environments validation."""
+
+    def extract_steps_environments(self):
```

Comment:
I believe this method can be private (`_extract_steps_environments`) as it's only used internally in this class.